{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A python recreation of the R script I used for my masters project, which:\n",
    "\n",
    "1) Analysed gene expression data from various Alloteropsis species and related donor species (in the context of lateral gene trasnfer)\n",
    "\n",
    "2) Filtered accessions by suitable orthogroups, data type as well.\n",
    "\n",
    "3) Performed statistical analysis on filtered datasets\n",
    "\n",
    "4) Performed suitable calculations to assess gene expression differences.\n",
    "\n",
    "5) Visualised data as appropriate.\n",
    "\n",
    "The goal of this project was to see whether laterally transferred genes (LGTs) were regulated differently from their native orthologues within the 'recipient' species, and/or the native orthologue from their donor 'gene'. Gene expression was used as a proxy for regulation. \n",
    "\n",
    "Information about species phylogeny was gathered prior to study, gene expression data was collected then processed via software available on Linux (with the use of shell scripting)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Importing packages for analysis\n",
    "import numpy as np # to convert data type\n",
    "import pandas as pd #pandas is a Python library (py terminology for package) whihc can help with data analysis, cleaning, manipulating and exploration\n",
    "import matplotlib.pyplot as plt #matplotlib is a library for creating various visualisations\n",
    "import seaborn as sns #assists in the creation of attractive and informative statistical graphs\n",
    "import os #provides functions for interacting with the operating system"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Filtering data:\n",
    "\n",
    "There is quite a lot of filters in the next section, I was time-strapped so chose to do things quite heavy handidly rather than shortening them with loops etc. but it got the job done. \n",
    "\n",
    "To understand the filters however an explanation is needed:\n",
    "\n",
    "Orthogroup: A number tag (data type: string) which identifies related genes across our study species (i.e. Orthogroup 1 contains phylogenetically related genes)\n",
    "\n",
    "Accessions: Geographically distinct accessions of Alloteropsis semialata (AUS, ZAM, KWT )\n",
    "Donors: Setaria italica (SET) and Themeda triandra (THE), two species identified as having donated genes to A. semialata\n",
    "Native (donor) genes: The vertically evolved native gene in our donor species.\n",
    "Native (recipient) genes: The vertically evolved native gene in our recipient species.\n",
    "LGTs: A gene thought have been laterally transferred to our A. semialata accessions (i.e. it's functional evolution occured in a different species)\n",
    "\n",
    "Correspondingly I named my dataframes to indicate the dataset we're filteirng for. E.g. AUS_LGT includes only LGTs from the AUS accession.... AUS_SET_native includes native genes Setaria italica that have LGT relatives in our AUS accession.\n",
    "\n",
    "The All_accessions data simply contains all of our Accession data (but not the SET/THE genes). These filters were done to allow for various different analysis. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Setting the working directory\n",
    "os.chdir(\"/home/joe/Desktop/Coding/Python/Masters_Project/expression_data\")\n",
    "os.getcwd() #check if wd properly set, note this doesn't output within this jupyter cell for some reason\n",
    "pd.set_option('display.max_columns', None)\n",
    "pd.set_option('display.max_rows', None)\n",
    "pd.options.mode.use_inf_as_na = False #set this to use tab separators for .csv files\n",
    "\n",
    "#Add some functions to reduce later commands \n",
    "orthogroup = list(range(1, 180)) #no of orthogroups set \n",
    "columns_to_convert = ['average root', 'average tip leaf', 'average overall', 'sd.root', 'se.root', 'sd.tip.leaf', 'se.tip.leaf']\n",
    "\n",
    "# Read the CSV file\n",
    "AUS_LGT = pd.read_csv(\"AUS_LGT.csv\")\n",
    "AUS_LGT['type'] = \"LGT\"\n",
    "AUS_native = pd.read_csv(\"AUS_native.csv\")\n",
    "AUS_SET_native = pd.read_csv(\"AUS_SET_native.csv\")\n",
    "AUS_THE_native = pd.read_csv(\"AUS_THE_native.csv\")\n",
    "AUS_native['type'] = \"Recipient\"\n",
    "AUS_SET_native['type'] = \"Donor\"\n",
    "AUS_THE_native['type'] = \"Donor\"\n",
    "\n",
    "#Make AUS list \n",
    "AUS_df = [AUS_LGT, AUS_native, AUS_SET_native, AUS_THE_native]\n",
    "for df in AUS_df:\n",
    "    df['orthogroup'] = orthogroup\n",
    "    df['orthogroup'] = df['orthogroup'].astype(str)\n",
    "    df[columns_to_convert] = df[columns_to_convert].apply(pd.to_numeric, errors='coerce') \n",
    "    df['accession'] = \"AUS\"\n",
    " \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read the CSV file\n",
    "ZAM_LGT = pd.read_csv(\"ZAM_LGT.csv\")\n",
    "ZAM_LGT['type'] = \"LGT\"\n",
    "ZAM_native = pd.read_csv(\"ZAM_native.csv\")\n",
    "ZAM_SET_native = pd.read_csv(\"ZAM_SET_native.csv\")\n",
    "ZAM_THE_native = pd.read_csv(\"ZAM_THE_native.csv\")\n",
    "ZAM_native['type'] = \"Recipient\"\n",
    "ZAM_SET_native['type'] = \"Donor\"\n",
    "ZAM_THE_native['type'] = \"Donor\"\n",
    "\n",
    "#Make ZAM list \n",
    "ZAM_df = [ZAM_LGT, ZAM_native, ZAM_SET_native, ZAM_THE_native]\n",
    "for df in ZAM_df:\n",
    "    df['orthogroup'] = orthogroup\n",
    "    df['orthogroup'] = df['orthogroup'].astype(str)\n",
    "    df[columns_to_convert] = df[columns_to_convert].apply(pd.to_numeric, errors='coerce') \n",
    "    df['accession'] = \"ZAM\"\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read the CSV file\n",
    "KWT_LGT = pd.read_csv(\"KWT_LGT.csv\")\n",
    "KWT_LGT['type'] = \"LGT\"\n",
    "KWT_native = pd.read_csv(\"KWT_native.csv\")\n",
    "KWT_SET_native = pd.read_csv(\"KWT_SET_native.csv\")\n",
    "KWT_THE_native = pd.read_csv(\"KWT_THE_native.csv\")\n",
    "KWT_native['type'] = \"Recipient\"\n",
    "KWT_SET_native['type'] = \"Donor\"\n",
    "KWT_THE_native['type'] = \"Donor\"\n",
    "\n",
    "#Make KWT list \n",
    "KWT_df = [KWT_LGT, KWT_native, KWT_SET_native, KWT_THE_native]\n",
    "for df in KWT_df:\n",
    "    df['orthogroup'] = orthogroup\n",
    "    df['orthogroup'] = df['orthogroup'].astype(str)\n",
    "    df[columns_to_convert] = df[columns_to_convert].apply(pd.to_numeric, errors='coerce') \n",
    "    df['accession'] = \"KWT\"\n",
    "    \n",
    "#OK all work as intended, remember to include proper indenting!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read CSV files for Setaria and Themeda data\n",
    "SET_native = pd.read_csv(\"All_SET_native.csv\")\n",
    "THE_native = pd.read_csv(\"All_THE_native.csv\")\n",
    "\n",
    "SET_native['orthogroup'] = orthogroup\n",
    "THE_native['orthogroup'] = orthogroup\n",
    "SET_native['orthogroup'] = SET_native['orthogroup'].astype(str)\n",
    "THE_native['orthogroup'] = THE_native['orthogroup'].astype(str)\n",
    "SET_native['type'] = \"SET\"\n",
    "THE_native['type'] = \"THE\"\n",
    "\n",
    "# Combine data from different dataframes into a single dataframe\n",
    "All_accessions = pd.concat([AUS_LGT, AUS_native, KWT_LGT, KWT_native, ZAM_LGT, ZAM_native])\n",
    "\n",
    "All_accessions = All_accessions[['gene', 'orthogroup', 'type', 'accession', 'average tip leaf', 'average root', 'donor']]\n",
    "All_accessions = All_accessions.dropna(subset=['average tip leaf'])\n",
    "# Check the column names of the All_accessions dataframe\n",
    "\n",
    "#Do the same for a df containing Donor genes too\n",
    "All_accessions_with_donor = pd.concat([AUS_LGT, AUS_native, KWT_LGT, KWT_native, ZAM_LGT, ZAM_native, SET_native, THE_native])\n",
    "\n",
    "# Select specific columns\n",
    "All_accessions_with_donor = All_accessions_with_donor[['gene', 'orthogroup', 'type', 'accession', 'average tip leaf', 'average root', 'donor']]\n",
    "\n",
    "# Filter rows where 'average.tip.leaf' is not NaN\n",
    "All_accessions_with_donor = All_accessions_with_donor.dropna(subset=['average tip leaf'])\n",
    "\n",
    "# Filter rows where 'donor' is either \"Cenchrinae\" or \"Andropogoneae\"\n",
    "All_accessions_with_donor = All_accessions_with_donor[All_accessions_with_donor['donor'].isin(['Cenchrinae', 'Andropogoneae'])]\n",
    "\n",
    "# print(All_accessions_with_donor['type'].unique()) seems to be good"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The next step is to filter our data to only contain:\n",
    "\n",
    "1) Genes linked to Cenchrinae or Andropogoneae donors (which we are using S. italica and T. triandra as proxies for)\n",
    "\n",
    "2) LGTs and Recipients within this dataset\n",
    "\n",
    "3) Orthogroups which have both data for LGTS and Recipient genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['170', '172', '39', '50', '163', '71', '76', '86', '89', '102', '106', '145', '148', '163', '170', '172', '174', '176', '177', '21', '26', '21', '117']\n"
     ]
    }
   ],
   "source": [
    " # Filter rows where donor is \"Cenchrinae\" or \"Andropogoneae\"\n",
    "All_accessions = All_accessions[(All_accessions['donor'] == \"Cenchrinae\") | (All_accessions['donor'] == \"Andropogoneae\")]\n",
    "\n",
    "# Filter rows where type is \"LGT\"\n",
    "All_accessions_LGT = All_accessions[All_accessions['type'] == \"LGT\"]\n",
    "\n",
    "# Filter rows where type is \"Recipient\"\n",
    "All_accessions_recipient = All_accessions[All_accessions['type'] == \"Recipient\"]\n",
    "\n",
    "# Perform anti-join to get rows where orthogroup is in LGT but not in Recipient\n",
    "LGTortho = All_accessions_LGT.merge(All_accessions_recipient, on=\"orthogroup\", how=\"left\", indicator=True)\n",
    "LGTortho = LGTortho[LGTortho['_merge'] == 'left_only']\n",
    "LGTortho.drop(columns='_merge', inplace=True)\n",
    "\n",
    "# Perform anti-join to get rows where orthogroup is in Recipient but not in LGT\n",
    "Recortho = All_accessions_recipient.merge(All_accessions_LGT, on=\"orthogroup\", how=\"left\", indicator=True)\n",
    "Recortho = Recortho[Recortho['_merge'] == 'left_only']\n",
    "Recortho.drop(columns='_merge', inplace=True)\n",
    "\n",
    "# Combine the results of LGTortho and Recortho\n",
    "Diffortho = pd.concat([LGTortho, Recortho])\n",
    "\n",
    "# List of orthogroups that aren't in both LGT and Recipient\n",
    "print(Diffortho['orthogroup'].tolist())\n",
    "\n",
    "# Filter rows where orthogroup is not in the specified list\n",
    "orthogroups_to_remove = [\"170\", \"172\", \"39\", \"50\", \"163\", \"71\", \"76\", \"86\", \"89\", \"102\", \"106\", \"145\", \"148\", \"163\", \"170\", \"172\", \"174\", \"176\", \"177\", \"21\", \"26\", \"21\", \"117\"]\n",
    "All_accessions = All_accessions[~All_accessions['orthogroup'].isin(orthogroups_to_remove)]\n",
    "\n",
    "#recreate the All_Accessions dataframes to include the orthogroup filtering\n",
    " # Filter rows where donor is \"Cenchrinae\" or \"Andropogoneae\"\n",
    "All_accessions = All_accessions[(All_accessions['donor'] == \"Cenchrinae\") | (All_accessions['donor'] == \"Andropogoneae\")]\n",
    "\n",
    "# Filter rows where type is \"LGT\"\n",
    "All_accessions_LGT = All_accessions[All_accessions['type'] == \"LGT\"]\n",
    "\n",
    "# Filter rows where type is \"Recipient\"\n",
    "All_accessions_recipient = All_accessions[All_accessions['type'] == \"Recipient\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now it's time for some statistical tests, the first is a Kruskal Wallis Chi-2 test:\n",
    "\n",
    "We want to see whether:\n",
    "\n",
    "1) Alloteropsis recipient gene expression differs between accessions (root and tip leaf)\n",
    "2) Alloteropsis LGT expression differs between accessions (root and tip leaf) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "KruskalResult(statistic=1.8117499837828739, pvalue=0.4041880690494699)\n",
      "KruskalResult(statistic=4.487708380395908, pvalue=0.10604898274925764)\n",
      "KruskalResult(statistic=1.7931571144700287, pvalue=0.40796309698494104)\n",
      "KruskalResult(statistic=0.5297991399224353, pvalue=0.7672830043683889)\n"
     ]
    }
   ],
   "source": [
    "from scipy.stats import stats #stats package\n",
    "from scipy.stats import kruskal \n",
    "#Recipient x root \n",
    "krr = kruskal(*[group['average root'].values for name, group in All_accessions_recipient.groupby('accession')])\n",
    "print(krr)\n",
    "#KruskalResult(statistic=1.8117499837828739, pvalue=0.4041880690494699)\n",
    "\n",
    "#Recipient x tip leaf\n",
    "krtl = kruskal(*[group['average tip leaf'].values for name, group in All_accessions_recipient.groupby('accession')])\n",
    "[print(krtl)]\n",
    "#KruskalResult(statistic=4.487708380395908, pvalue=0.10604898274925764)\n",
    "\n",
    "#LGT x root\n",
    "klr = kruskal(*[group['average root'].values for name, group in All_accessions_LGT.groupby('accession')])\n",
    "print(klr)\n",
    "#KruskalResult(statistic=1.7931571144700287, pvalue=0.40796309698494104)\n",
    "\n",
    "#LGT x tip leaf \n",
    "kltl = kruskal(*[group['average tip leaf'].values for name, group in All_accessions_LGT.groupby('accession')])\n",
    "print(kltl)\n",
    "#KruskalResult(statistic=0.5297991399224353, pvalue=0.7672830043683889)\n",
    "\n",
    "#Slight variation explained by differences in python/R algorithms \n",
    "\n",
    "#Okay... we finally sorted out our data and got the tests done successfully...\n",
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now for the next step... the recreation of Figure 1...\n",
    "\n",
    "This was made to see the distribution of gene expression across our accessions. There should a histogram for the tip leaf and root expression, as well as replicate graphs using a narrower range on the X axis (for higher clarity)\n",
    "\n",
    "NOTE: I had EXTREME trouble trying to replicate this plot in Python, I resorted to using the plotnine (ggplot) package to get an output that was the same as 'R'. I believe Python by default must produce its histograms differently than ggplot2 does, and even ggplot2 required significant tinkering and configuration to get it to look the same...\n",
    "\n",
    "Additionally, I left the colour defaults as they were deliberately to distinguish the ones made on python vs R.\n",
    "\n",
    "My other attempt using seaborn was probably successful, but the visualisation differed for some unknown reason... I even checked all the data against my original R script but there just seemed to be missing data. I tried changing the binwidth, tried deliberately including '0' values but couldn't find any explanations on why the data output differed. \n",
    "\n",
    "Also ancedotally, it seems python just adjusts the colours sometimes when using the exact same code. I also feel 'R' is more prone to errors on Linux relative to Windows..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  accession  average root\n",
      "0       AUS    247.136528\n",
      "1       KWT    218.920583\n",
      "2       ZAM    812.320797\n"
     ]
    }
   ],
   "source": [
    "# Does All_accessions actually match between 'R' and 'Python'... yes is the answer. \n",
    "\n",
    "# Filter the DataFrame to include rows where 'average.root' is between 0 and 30\n",
    "filtered_accessions = All_accessions[(All_accessions['average root'] >= 0) & (All_accessions['average root'] <= 30)]\n",
    "\n",
    "# Group the data by 'accession' and calculate the sum of 'average.root' for each group\n",
    "accession_sums = filtered_accessions.groupby('accession')['average root'].sum().reset_index()\n",
    "\n",
    "print(accession_sums)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'ggplot' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_46533/2267357513.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m# Plot the histogram for ROOT expression\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mplotROOT\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mggplot\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mAll_accessions\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maes\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'average root'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfill\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'accession'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m+\u001b[0m\u001b[0;31m \u001b[0m\u001b[0;31m\\\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m        \u001b[0mgeom_histogram\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mbinwidth\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m30\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m+\u001b[0m\u001b[0;31m \u001b[0m\u001b[0;31m\\\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m        \u001b[0mfacet_wrap\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'~type'\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m+\u001b[0m\u001b[0;31m \u001b[0m\u001b[0;31m\\\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m        \u001b[0mxlab\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Average root expression\"\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m+\u001b[0m\u001b[0;31m \u001b[0m\u001b[0;31m\\\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'ggplot' is not defined"
     ]
    }
   ],
   "source": [
    "# Plot the histogram for ROOT expression \n",
    "plotROOT = ggplot(All_accessions, aes(x='average root', fill='accession')) + \\\n",
    "       geom_histogram(binwidth=30) + \\\n",
    "       facet_wrap('~type') + \\\n",
    "       xlab(\"Average root expression\") + \\\n",
    "       labs(fill=\"Accession\")\n",
    "\n",
    "# Save the plot to a file with width and height in inches\n",
    "ggsave(plotROOT, filename=\"/home/joe/Desktop/Coding/Python/Masters_Project/graphs/results1/plotnine_root_histogram_all.png\", width=12, height=7)\n",
    "plotROOT\n",
    "\n",
    "plotROOT250 = ggplot(All_accessions, aes(x='average root', fill='accession')) + \\\n",
    "       geom_histogram(binwidth=10) + \\\n",
    "       facet_wrap('~type') + \\\n",
    "       xlab(\"Average root expression\") + \\\n",
    "       labs(fill=\"Accession\") + \\\n",
    "       scale_x_continuous(limits=(0, 250)) + \\\n",
    "       scale_y_continuous(limits=(0, 30))\n",
    "\n",
    "# Save the plot to a file with width and height in inches\n",
    "ggsave(plotROOT250, filename=\"/home/joe/Desktop/Coding/Python/Masters_Project/graphs/results1/plotnine_root_histogram_250.png\", width=12, height=7)\n",
    "\n",
    "# Display the plot\n",
    "print(plotROOT250)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Does All_accessions actually match between 'R' and 'Python'... yes is the answer. \n",
    "\n",
    "# Filter the DataFrame to include rows where 'average.tip leaf' is between 0 and 30\n",
    "filtered_accessions = All_accessions[(All_accessions['average tip leaf'] >= 0) & (All_accessions['average tip leaf'] <= 30)]\n",
    "\n",
    "# Group the data by 'accession' and calculate the sum of 'average.root' for each group\n",
    "accession_sums = filtered_accessions.groupby('accession')['average tip leaf'].sum().reset_index()\n",
    "\n",
    "print(accession_sums)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot the histogram for tip leaf expression \n",
    "plotLEAF = ggplot(All_accessions, aes(x='average tip leaf', fill='accession')) + \\\n",
    "       geom_histogram(binwidth=300) + \\\n",
    "       facet_wrap('~type') + \\\n",
    "       xlab(\"Average tip leaf expression\") + \\\n",
    "       labs(fill=\"Accession\")\n",
    "\n",
    "# Save the plot to a file with width and height in inches\n",
    "ggsave(plotLEAF, filename=\"/home/joe/Desktop/Coding/Python/Masters_Project/graphs/results1/plotnine_tip_leaf_histogram_all.png\", width=12, height=7)\n",
    "\n",
    "plotLEAF \n",
    "\n",
    "plotLEAF250 = ggplot(All_accessions, aes(x='average tip leaf', fill='accession')) + \\\n",
    "       geom_histogram(binwidth=10) + \\\n",
    "       facet_wrap('~type') + \\\n",
    "       xlab(\"Average tip leaf expression\") + \\\n",
    "       labs(fill=\"Accession\") + \\\n",
    "       scale_x_continuous(limits=(0, 250)) + \\\n",
    "       scale_y_continuous(limits=(0, 30))\n",
    "       \n",
    "plotLEAF250\n",
    "ggsave(plotLEAF, filename=\"/home/joe/Desktop/Coding/Python/Masters_Project/graphs/results1/plotnine_tip_leaf_histogram_250.png\", width=12, height=7)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Histograms done and saved... now I need to:\n",
    "\n",
    "1) Make appropriate filters for Figure 2, which looks at the relative expression of shared 'native recipient' genes in KWT vs AUS, KWT vs ZAM and ZAM vs AUS for both the roots and leaves respectively. \n",
    "\n",
    "2) Calculate average ranges for purposes of plotting.\n",
    "\n",
    "3) Create Figure 2 plots \n",
    "\n",
    "4) Create Figure 3 plot, which is a comparison of root x leaf expression for LGTs, (all) native recipients, SET native and THE native. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dotplotleaf = All_accessions[['average tip leaf', 'orthogroup', 'type', 'accession']].copy()\n",
    "dotplotleaf.rename(columns={'average tip leaf': 'expression'}, inplace=True)\n",
    "dotplotleaf['area'] = \"Tip.leaf\"\n",
    "\n",
    "dotplotroot = All_accessions[['average root', 'orthogroup', 'type', 'accession']].copy()\n",
    "dotplotroot.rename(columns={'average root': 'expression'}, inplace=True)\n",
    "dotplotroot['area'] = \"Root\"\n",
    "\n",
    "dotplot = pd.concat([dotplotleaf, dotplotroot], ignore_index=True)\n",
    "\n",
    "dotplotZAM = dotplot[dotplot['accession'] == \"ZAM\"].copy()\n",
    "dotplotZAM.rename(columns={'expression': 'ZAM_expression'}, inplace=True)\n",
    "dotplotZAM = dotplotZAM[['ZAM_expression', 'orthogroup', 'type', 'area']]\n",
    "\n",
    "dotplotKWT = dotplot[dotplot['accession'] == \"KWT\"].copy()\n",
    "dotplotKWT.rename(columns={'expression': 'KWT_expression'}, inplace=True)\n",
    "dotplotKWT = dotplotKWT[['KWT_expression', 'orthogroup', 'type', 'area']]\n",
    "\n",
    "dotplotAUS = dotplot[dotplot['accession'] == \"AUS\"].copy()\n",
    "dotplotAUS.rename(columns={'expression': 'AUS_expression'}, inplace=True)\n",
    "dotplotAUS = dotplotAUS[['AUS_expression', 'orthogroup', 'type', 'area']]\n",
    "\n",
    "dotplotAUSKWT = pd.merge(dotplotAUS, dotplotKWT)\n",
    "dotplotAUSZAM = pd.merge(dotplotZAM, dotplotAUS)\n",
    "dotplotKWTZAM = pd.merge(dotplotZAM, dotplotKWT)\n",
    "\n",
    "dotplotAUSKWT['difference'] = dotplotAUSKWT['AUS_expression'] - dotplotAUSKWT['KWT_expression']\n",
    "dotplotAUSZAM['difference'] = dotplotAUSZAM['AUS_expression'] - dotplotAUSZAM['ZAM_expression']\n",
    "dotplotKWTZAM['difference'] = dotplotKWTZAM['KWT_expression'] - dotplotKWTZAM['ZAM_expression']\n",
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Group by 'area' and 'type', then calculate the IQR of 'AUS_expression' for both AUS and KWT \n",
    "dotplotAUSKWT_IQR_AUS = dotplotAUSKWT.groupby(['area', 'type'])['AUS_expression'].quantile(0.75) - dotplotAUSKWT.groupby(['area', 'type'])['AUS_expression'].quantile(0.25)\n",
    "dotplotAUSKWT_IQR_AUS = dotplotAUSKWT_IQR_AUS.reset_index() # Reset the index to have a nice tabular format\n",
    "dotplotAUSKWT_IQR_AUS.rename(columns={'AUS_expression': 'AUS_expression_IQR'}, inplace=True) #rename\n",
    "#print(dotplotAUSKWT_IQR_AUS) #print results\n",
    "\n",
    "#Now for KWT \n",
    "dotplotAUSKWT_IQR_KWT = dotplotAUSKWT.groupby(['area', 'type'])['KWT_expression'].quantile(0.75) - dotplotAUSKWT.groupby(['area', 'type'])['KWT_expression'].quantile(0.25)\n",
    "dotplotAUSKWT_IQR_KWT = dotplotAUSKWT_IQR_KWT.reset_index()\n",
    "dotplotAUSKWT_IQR_KWT.rename(columns={'KWT_expression': 'KWT_expression_IQR'}, inplace=True)\n",
    "\n",
    "#Now for AUSZAM datasets, starting with AUS\n",
    "dotplotAUSZAM_IQR_AUS = dotplotAUSZAM.groupby(['area', 'type'])['AUS_expression'].quantile(0.75) - dotplotAUSZAM.groupby(['area', 'type'])['AUS_expression'].quantile(0.25)\n",
    "dotplotAUSZAM_IQR_AUS = dotplotAUSZAM_IQR_AUS.reset_index() # Reset the index to have a nice tabular format\n",
    "dotplotAUSZAM_IQR_AUS.rename(columns={'AUS_expression': 'AUS_expression_IQR'}, inplace=True) #rename\n",
    "\n",
    "#ZAM\n",
    "dotplotAUSZAM_IQR_ZAM = dotplotAUSZAM.groupby(['area', 'type'])['ZAM_expression'].quantile(0.75) - dotplotAUSZAM.groupby(['area', 'type'])['ZAM_expression'].quantile(0.25)\n",
    "dotplotAUSZAM_IQR_ZAM = dotplotAUSZAM_IQR_ZAM.reset_index() # Reset the index to have a nice tabular format\n",
    "dotplotAUSZAM_IQR_ZAM.rename(columns={'ZAM_expression': 'ZAM_expression_IQR'}, inplace=True) #rename\n",
    "\n",
    "#Now for ZAMKWT (KWT first)\n",
    "dotplotKWTZAM_IQR_KWT = dotplotKWTZAM.groupby(['area', 'type'])['KWT_expression'].quantile(0.75) - dotplotKWTZAM.groupby(['area', 'type'])['KWT_expression'].quantile(0.25)\n",
    "dotplotKWTZAM_IQR_KWT = dotplotKWTZAM_IQR_KWT.reset_index() # Reset the index to have a nice tabular format\n",
    "dotplotKWTZAM_IQR_KWT.rename(columns={'KWT_expression': 'KWT_expression_IQR'}, inplace=True) #rename\n",
    "print(dotplotKWTZAM_IQR_KWT)\n",
    "\n",
    "#Now for ZAMKWT (ZAM)\n",
    "dotplotKWTZAM_IQR_ZAM = dotplotKWTZAM.groupby(['area', 'type'])['ZAM_expression'].quantile(0.75) - dotplotKWTZAM.groupby(['area', 'type'])['ZAM_expression'].quantile(0.25)\n",
    "dotplotKWTZAM_IQR_ZAM = dotplotKWTZAM_IQR_ZAM.reset_index() # Reset the index to have a nice tabular format\n",
    "dotplotKWTZAM_IQR_ZAM.rename(columns={'ZAM_expression': 'ZAM_expression_IQR'}, inplace=True) #rename\n",
    "print(dotplotKWTZAM_IQR_ZAM)\n",
    "\n",
    "#For genes contained in multiple accessions \n",
    "root_IQR = All_accessions_with_donor.groupby(['accession', 'type'])['average root'].agg(lambda x: np.percentile(x, 75) - np.percentile(x, 25))\n",
    "tip_leaf_IQR = All_accessions_with_donor.groupby(['accession', 'type'])['average tip leaf'].agg(lambda x: np.percentile(x, 75) - np.percentile(x, 25))\n",
    "#print(root_IQR)\n",
    "#print(tip_leaf_IQR)\n",
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#KWT x AUS dotplot \n",
    "# Set the style to \"whitegrid\" for a distinct background\n",
    "sns.set_style(\"whitegrid\")\n",
    "\n",
    "# Plot the scatter plot with seaborn\n",
    "sns.scatterplot(data=dotplotAUSKWT, x='AUS_expression', y='KWT_expression', hue='area', style='type')\n",
    "\n",
    "# Add the black diagonal line\n",
    "plt.plot([0, 70], [0, 70], color='black')\n",
    "\n",
    "# Set the x and y-axis limits\n",
    "plt.xlim(0, 70)\n",
    "plt.ylim(0, 70)\n",
    "\n",
    "# Set the axis labels\n",
    "plt.xlabel('AUS_expression')\n",
    "plt.ylabel('KWT_expression')\n",
    "\n",
    "# Show the legend\n",
    "plt.legend(title='Area')\n",
    "\n",
    "# Create the directory if it doesn't exist\n",
    "output_dir = \"/home/joe/Desktop/Coding/Python/Masters_Project/graphs/results2/\"\n",
    "os.makedirs(output_dir, exist_ok=True)\n",
    "\n",
    "# Save the plot to the specified directory\n",
    "output_filename = \"AUSxKWT_sb.png\"\n",
    "output_path = os.path.join(output_dir, output_filename)\n",
    "plt.savefig(output_path)\n",
    "\n",
    "# Show the plot\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#ZAM x AUS dotplot \n",
    "# Set the style to \"whitegrid\" for a distinct background\n",
    "sns.set_style(\"whitegrid\")\n",
    "\n",
    "# Plot the scatter plot with seaborn\n",
    "sns.scatterplot(data=dotplotAUSZAM, x='AUS_expression', y='ZAM_expression', hue='area', style='type')\n",
    "\n",
    "# Add the black diagonal line\n",
    "plt.plot([0, 70], [0, 70], color='black')\n",
    "\n",
    "# Set the x and y-axis limits\n",
    "plt.xlim(0, 70)\n",
    "plt.ylim(0, 70)\n",
    "\n",
    "# Set the axis labels\n",
    "plt.xlabel('AUS_expression')\n",
    "plt.ylabel('ZAM_expression')\n",
    "\n",
    "# Create the directory if it doesn't exist\n",
    "output_dir = \"/home/joe/Desktop/Coding/Python/Masters_Project/graphs/results2/\"\n",
    "os.makedirs(output_dir, exist_ok=True)\n",
    "\n",
    "# Save the plot to the specified directory\n",
    "output_filename = \"AUSxZAM_sb.png\"\n",
    "output_path = os.path.join(output_dir, output_filename)\n",
    "plt.savefig(output_path)\n",
    "\n",
    "# Show the plot\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#ZAM x AUS dotplot \n",
    "# Set the style to \"whitegrid\" for a distinct background\n",
    "sns.set_style(\"whitegrid\")\n",
    "\n",
    "# Plot the scatter plot with seaborn\n",
    "sns.scatterplot(data=dotplotKWTZAM, x='KWT_expression', y='ZAM_expression', hue='area', style='type')\n",
    "\n",
    "# Add the black diagonal line\n",
    "plt.plot([0, 70], [0, 70], color='black')\n",
    "\n",
    "# Set the x and y-axis limits\n",
    "plt.xlim(0, 70)\n",
    "plt.ylim(0, 70)\n",
    "\n",
    "# Set the axis labels\n",
    "plt.xlabel('KWT_expression')\n",
    "plt.ylabel('ZAM_expression')\n",
    "\n",
    "# Create the directory if it doesn't exist\n",
    "output_dir = \"/home/joe/Desktop/Coding/Python/Masters_Project/graphs/results2/\"\n",
    "os.makedirs(output_dir, exist_ok=True)\n",
    "\n",
    "# Save the plot to the specified directory\n",
    "output_filename = \"KWTxZAM_sb.png\"\n",
    "output_path = os.path.join(output_dir, output_filename)\n",
    "plt.savefig(output_path)\n",
    "\n",
    "# Show the plot\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Now for Figure 3 \n",
    "# Set the style to \"whitegrid\" for a distinct background\n",
    "sns.set_style(\"whitegrid\")\n",
    "\n",
    "# Plot the scatter plot with seaborn\n",
    "sns.scatterplot(data=All_accessions_with_donor, x='average tip leaf', y='average root', hue='type')\n",
    "\n",
    "# Add the black diagonal line\n",
    "plt.plot([0, 250], [0, 250], color='black')\n",
    "\n",
    "# Set the x and y-axis limits\n",
    "plt.xlim(0, 250)\n",
    "plt.ylim(0, 250)\n",
    "\n",
    "# Set the axis labels\n",
    "plt.xlabel('Average tip leaf expression')\n",
    "plt.ylabel('Average root expression')\n",
    "\n",
    "# Create the directory if it doesn't exist\n",
    "output_dir = \"/home/joe/Desktop/Coding/Python/Masters_Project/graphs/results3/\"\n",
    "os.makedirs(output_dir, exist_ok=True)\n",
    "\n",
    "# Save the plot to the specified directory\n",
    "output_filename = \"RootvsTipleaf.png\"\n",
    "output_path = os.path.join(output_dir, output_filename)\n",
    "plt.savefig(output_path)\n",
    "\n",
    "# Show the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now for some calculations/stats to check:\n",
    "\n",
    "1) Whether expression was significantly different between the root and leaf tip for:\n",
    "    - Each donor species native genes (SET and THE)\n",
    "    - Recipient (Alloteropsis) native genes\n",
    "    - LGTs identified in Alloteropsis \n",
    "2) If there was a correlation between tip leaf and root expression for:\n",
    "    - Recipient genes\n",
    "    - LGTs \n",
    "    - SET and THE native genes\n",
    "3) The IQR of root and tip leaf expression (for the purposes of tables) for:\n",
    "    - THE and SET\n",
    "    - LGTs\n",
    "    - Recipient genes "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "61\n"
     ]
    }
   ],
   "source": [
    "# Select columns and rename them\n",
    "from scipy.stats import wilcoxon #\n",
    "\n",
    "root_wilcox = All_accessions[['orthogroup', 'type', 'average root', 'donor']]\n",
    "root_wilcox = root_wilcox.rename(columns={'average root': 'expression'})\n",
    "root_wilcox['area'] = 'root'\n",
    "\n",
    "tip_leaf_wilcox = All_accessions[['orthogroup', 'type', 'average tip leaf', 'donor']]\n",
    "tip_leaf_wilcox = tip_leaf_wilcox.rename(columns={'average tip leaf': 'expression'})\n",
    "tip_leaf_wilcox['area'] = 'tip leaf'\n",
    "\n",
    "wilcox_root_leaf = pd.concat([root_wilcox, tip_leaf_wilcox], ignore_index=True)\n",
    "#print(wilcox_root_leaf) \n",
    "#data successfully filtered to include 'expression' and 'area' separately\n",
    "\n",
    "# Now to look at LGT and recipients specifically\n",
    "\n",
    "wilcox_root_leaf_LGT = wilcox_root_leaf[wilcox_root_leaf['type'] == 'LGT']\n",
    "wilcox_root_leaf_Recipient = wilcox_root_leaf[wilcox_root_leaf['type'] == 'Recipient']\n",
    "\n",
    "print((wilcox_root_leaf_LGT['expression'] == 0).sum()) #checking matching data \n",
    "\n",
    "#LGT comparisons with THE and SET native genes\n",
    "wilcoxrootleafLGT_THE = wilcox_root_leaf_LGT[wilcox_root_leaf_LGT['donor'] == \"Andropogoneae\"]\n",
    "wilcoxrootleafLGT_SET = wilcox_root_leaf_LGT[wilcox_root_leaf_LGT['donor'] == \"Cenchrinae\"]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'SET_wilcox_root'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_46533/3302682243.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     15\u001b[0m         \u001b[0;31m# Set new dataframe name\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     16\u001b[0m         \u001b[0mnew_name\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34mf\"{names[i]}_{region.replace(' ', '_')}\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 17\u001b[0;31m         \u001b[0mglobals\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mnew_name\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'expression'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto_numeric\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mglobals\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mnew_name\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'expression'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     18\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     19\u001b[0m \u001b[0;31m#loop to make SET_wilcox_root and SET_wilcox_tip leaf... not needed butI just wanted\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'SET_wilcox_root'"
     ]
    }
   ],
   "source": [
    "#Native root vs tip leaf comparions for THE and SET\n",
    "SET_wilcox_native = All_accessions_with_donor[All_accessions_with_donor['type'] == \"SET\"]\n",
    "THE_wilcox_native = All_accessions_with_donor[All_accessions_with_donor['type'] == \"THE\"]\n",
    "\n",
    "root_tipleaf = [\"root\", \"tip leaf\"]\n",
    "dataframes = [SET_wilcox_native, THE_wilcox_native]\n",
    "names = ['SET_wilcox', 'THE_wilcox']\n",
    "\n",
    "for i, df in enumerate(dataframes):\n",
    "    for region in root_tipleaf:\n",
    "        processed_df = df[['orthogroup', 'type', f'average {region}']].copy()\n",
    "        processed_df.rename(columns={f'average {region}': 'expression'}, inplace=True)\n",
    "        processed_df['area'] = region\n",
    "\n",
    "        # Set new dataframe name\n",
    "        new_name = f\"{names[i]}_{region.replace(' ', '_')}\"\n",
    "        globals()[new_name]['expression'] = pd.to_numeric(globals()[new_name]['expression'])\n",
    "        \n",
    "#loop to make SET_wilcox_root and SET_wilcox_tip leaf... not needed butI just wanted\n",
    "#to do it for future reference so I don't have such long scripts...\n",
    "SET_wilcox_tiproot = pd.concat([SET_wilcox_root, SET_wilcox_tip_leaf], ignore_index=True)\n",
    "THE_wilcox_tiproot = pd.concat([THE_wilcox_root, THE_wilcox_tip_leaf], ignore_index=True)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now for some tests... but there's something interesting to consider that I hadn't before and I've noticed while all significant, my P-values can match or not match my 'R' script dependant on some settings.\n",
    "\n",
    "I noticed this because my R values for \"wilcox_root_leaf_LGT\" are V = 3592, p-value = 0.002653 \n",
    "\n",
    "1) Continuity correction (correction = True/False)... this generally is set to False for datasets larger than 50. It has a very minor impact on my p value. \n",
    "\n",
    "2) Zero_method = 'zsplit', zsplit handles ties by separating them into separate ranks for each half... it's apparently more accurate even when 61/423 values are zero. It reduces my p value from  ~0.00265 -> ~0.00139 which doesn't yield any new information on significance. \n",
    "\n",
    "3) Alternative = 'two-sided' this is the default option, to make it a two-tailed test and not specify a directional difference (i.e. some values may be higher, some lower) which is suitable for looking for DIFFERENCE rather than higher or lower expression in general.\n",
    "\n",
    "Interestingly... R seems to default to Correlation = 'True', Zero_method = 'Wilcox', 'Alternative = 'two-sided', and uses 'V' as a test statstic, which python DOES NOT calculate.\n",
    "\n",
    "V is the minimum out of W+ and W- (the alternative = greater/less test statistic) and is designed to provide a more accurate representation of the test statistic while accounting for tied values of data... so we can just take the minimum one of these and use it as V if needed.\n",
    "\n",
    "Regardless of the confusion... it is over now... I also decided to do this all as a loop unlike R. Note the indents have to match even on the prints, the added print() at the end ensures a space between results\n",
    "  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#wilcox_root_leaf_Recipient tests, efficient edition\n",
    "wilcox_list =  [wilcox_root_leaf_LGT, wilcox_root_leaf_Recipient, wilcoxrootleafLGT_THE, wilcoxrootleafLGT_SET, SET_wilcox_tiproot, THE_wilcox_tiproot]\n",
    "wilcox_names = [\"Wilcox test root/leaf LGT\", \"Wilcox test root/leaf Recipient\", \"Wilcox test root/leaf LGT THE\", \"Wilcox test root/leaf LGT SET\", \"Wilcox test root/leaf SET native\", \"Wilcox test root/leaf THE native\"]\n",
    "\n",
    "for i in range(len(wilcox_list)):\n",
    "    wilcoxtest = wilcox_list[i]\n",
    "    results = wilcoxon(wilcoxtest[wilcoxtest['area'] == 'root']['expression'],\n",
    "                                wilcoxtest[wilcoxtest['area'] == 'tip leaf']['expression'],\n",
    "                                alternative='two-sided',\n",
    "                                zero_method='wilcox',\n",
    "                                correction=True)\n",
    "\n",
    "    resultsg = wilcoxon(wilcoxtest[wilcoxtest['area'] == 'root']['expression'],\n",
    "                                wilcoxtest[wilcoxtest['area'] == 'tip leaf']['expression'],\n",
    "                                alternative='greater',\n",
    "                                zero_method='wilcox',\n",
    "                                correction=True)\n",
    "\n",
    "\n",
    "    resultsl = wilcoxon(wilcoxtest[wilcoxtest['area'] == 'root']['expression'],\n",
    "                                wilcoxtest[wilcoxtest['area'] == 'tip leaf']['expression'],\n",
    "                                alternative='less',\n",
    "                                zero_method='wilcox',\n",
    "                                correction=True)\n",
    "\n",
    "\n",
    "    W_plus = resultsg.statistic \n",
    "    W_minus = resultsl.statistic\n",
    "    V_statistic = min(W_plus, W_minus)\n",
    "    print(f\"Results for {wilcox_names[i]}:\") #the f-string allows us to use {wilcox_names[i]} as a loop\n",
    "    print(\"P-value =\",results.pvalue)\n",
    "    print(\"V Statistic =\", V_statistic)\n",
    "    print()\n",
    "#output matches R code \n",
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# LGT correlations - test statistic differs from 'R' but rest is fine \n",
    "\n",
    "from scipy.stats import spearmanr\n",
    "tip_leaf_lgt = All_accessions_LGT['average tip leaf']\n",
    "root_lgt = All_accessions_LGT['average root']\n",
    "spearman_corr_lgt, p_value_lgt = spearmanr(tip_leaf_lgt, root_lgt)\n",
    "\n",
    "print(\"Correlation results for LGT root tip leaf:\")\n",
    "print(\"Test Statistic =\", spearman_corr_lgt)\n",
    "print(\"p-value =\", p_value_lgt)\n",
    "print(\"rho =\", spearman_corr_lgt)\n",
    "print(\"R2 =\", spearman_corr_lgt**2)\n",
    "print()\n",
    "\n",
    "# Recipient correlations\n",
    "tip_leaf_recipient = All_accessions_recipient['average tip leaf']\n",
    "root_recipient = All_accessions_recipient['average root']\n",
    "spearman_corr_recipient, p_value_recipient = spearmanr(tip_leaf_recipient, root_recipient)\n",
    "\n",
    "print(\"Correlation results for Recipient root tip leaf:\")\n",
    "print(\"Test Statistic =\", spearman_corr_recipient)\n",
    "print(\"p-value =\", p_value_recipient)\n",
    "print(\"rho =\", spearman_corr_recipient)\n",
    "print(\"R2 =\", spearman_corr_recipient**2)\n",
    "print()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now time to get some IQRS "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Index: 43 entries, 15 to 177\n",
      "Data columns (total 7 columns):\n",
      " #   Column            Non-Null Count  Dtype \n",
      "---  ------            --------------  ----- \n",
      " 0   gene              43 non-null     object\n",
      " 1   orthogroup        43 non-null     object\n",
      " 2   type              43 non-null     object\n",
      " 3   accession         0 non-null      object\n",
      " 4   average tip leaf  43 non-null     object\n",
      " 5   average root      43 non-null     object\n",
      " 6   donor             43 non-null     object\n",
      "dtypes: object(7)\n",
      "memory usage: 2.7+ KB\n",
      "None\n"
     ]
    }
   ],
   "source": [
    "print(THE_wilcox_native.info())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "THE - Root IQR:\n",
      "type\n",
      "THE    33.75475\n",
      "Name: average root, dtype: float64\n",
      "\n",
      "THE - Tip Leaf IQR:\n",
      "type\n",
      "THE    21.04\n",
      "Name: average tip leaf, dtype: float64\n",
      "\n",
      "SET - Root IQR:\n",
      "type\n",
      "SET    26.62\n",
      "Name: average root, dtype: float64\n",
      "\n",
      "SET - Tip Leaf IQR:\n",
      "type\n",
      "SET    23.271111\n",
      "Name: average tip leaf, dtype: float64\n",
      "\n",
      "\n",
      "All Accessions - Root IQR:\n",
      "type\n",
      "LGT          16.149729\n",
      "Recipient    31.265833\n",
      "Name: average root, dtype: float64\n",
      "\n",
      "All Accessions - Tip Leaf IQR:\n",
      "type\n",
      "LGT          10.451167\n",
      "Recipient    21.558972\n",
      "Name: average tip leaf, dtype: float64\n"
     ]
    }
   ],
   "source": [
    "#THE IQR native\n",
    "THE_wilcox_native['average root'] = pd.to_numeric(THE_wilcox_native['average root'], errors='coerce') # Convert 'average root' column to numeric data type\n",
    "THE_wilcox_root_iqr = THE_wilcox_native.groupby('type')['average root'].quantile(0.75) - THE_wilcox_native.groupby('type')['average root'].quantile(0.25) # Calculate Root IQR\n",
    "THE_wilcox_native['average tip leaf'] = pd.to_numeric(THE_wilcox_native['average tip leaf'], errors='coerce') # Convert 'average tip leaf' column to numeric data type\n",
    "THE_wilcox_tip_leaf_iqr = THE_wilcox_native.groupby('type')['average tip leaf'].quantile(0.75) - THE_wilcox_native.groupby('type')['average tip leaf'].quantile(0.25) # Calculate Tip Leaf IQR\n",
    "\n",
    "print(\"THE - Root IQR:\")\n",
    "print(THE_wilcox_root_iqr)\n",
    "print()\n",
    "print(\"THE - Tip Leaf IQR:\")\n",
    "print(THE_wilcox_tip_leaf_iqr)\n",
    "print()\n",
    "\n",
    "#SET IQR \n",
    "SET_wilcox_native['average root'] = pd.to_numeric(SET_wilcox_native['average root'], errors='coerce') # Convert 'average root' column to numeric data type\n",
    "SET_wilcox_root_iqr = SET_wilcox_native.groupby('type')['average root'].quantile(0.75) - SET_wilcox_native.groupby('type')['average root'].quantile(0.25) # Calculate Root IQR\n",
    "SET_wilcox_native['average tip leaf'] = pd.to_numeric(SET_wilcox_native['average tip leaf'], errors='coerce') # Convert 'average root' column to numeric data type\n",
    "SET_wilcox_tip_leaf_iqr = SET_wilcox_native.groupby('type')['average tip leaf'].quantile(0.75) - SET_wilcox_native.groupby('type')['average tip leaf'].quantile(0.25) # Calculate Tip Leaf IQR\n",
    "\n",
    "print(\"SET - Root IQR:\")\n",
    "print(SET_wilcox_root_iqr)\n",
    "print()\n",
    "print(\"SET - Tip Leaf IQR:\")\n",
    "print(SET_wilcox_tip_leaf_iqr)\n",
    "print()\n",
    "\n",
    "# All accessions\n",
    "All_accessions_tip_leaf_iqr = All_accessions.groupby('type')['average tip leaf'].quantile(0.75) - All_accessions.groupby('type')['average tip leaf'].quantile(0.25)\n",
    "All_accessions_root_iqr = All_accessions.groupby('type')['average root'].quantile(0.75) - All_accessions.groupby('type')['average root'].quantile(0.25)\n",
    "\n",
    "print(\"\\nAll Accessions - Root IQR:\")\n",
    "print(All_accessions_root_iqr)\n",
    "\n",
    "print(\"\\nAll Accessions - Tip Leaf IQR:\")\n",
    "print(All_accessions_tip_leaf_iqr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "All Accessions - Root IQR:\n",
      "type\n",
      "LGT          16.149729\n",
      "Recipient    31.265833\n",
      "Name: average root, dtype: float64\n",
      "\n",
      "All Accessions - Tip Leaf IQR:\n",
      "type\n",
      "LGT          10.451167\n",
      "Recipient    21.558972\n",
      "Name: average tip leaf, dtype: float64\n"
     ]
    }
   ],
   "source": [
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
